src/ota_proxy/ota_cache.py

```diff
@@ -542,38 +544,42 @@ async def retrieve_file(
             return _res
 
         # --- case 4: no cache available, streaming remote file and cache --- #
-        tracker, is_writer = await self._on_going_caching.get_tracker(
-            cache_identifier,
+        # a online tracker is available for this requrest
+        if (tracker := self._on_going_caching.get_tracker(cache_identifier)) and (
+            subscription := await tracker.subscribe_tracker()
+        ):
+            # logger.debug(f"reader subscribe for {tracker.meta=}")
+            stream_fd, cache_meta = subscription
+            return stream_fd, cache_meta.export_headers_to_client()
+
+        # no valid online tracker is available for this request, create a new one and
+        #   promote the caller to be the provider.
+        # NOTE: register the tracker before open the remote fd!
+        tracker = CacheTracker(
+            cache_identifier=cache_identifier,
+            base_dir=self._base_dir,
             executor=self._executor,
-            callback=self._commit_cache_callback,
+            commit_cache_cb=self._commit_cache_callback,
             below_hard_limit_event=self._storage_below_hard_limit_event,
         )
-        if is_writer:
-            try:
-                remote_fd, resp_headers = await self._retrieve_file_by_downloading(
-                    raw_url, headers=headers_from_client
-                )
-            except Exception:
-                await tracker.provider_on_failed()
-                raise
+        self._on_going_caching.register_tracker(cache_identifier, tracker)
```

Comment:
Not handing over the event loop here if the caller is provider, immediately create a tracker and register it right after get_tracker failed.